tools/commands/build.py

```diff
@@ -66,40 +87,103 @@ def make(self):
+        self.make_verbose = make_verbose
+        self.host_prefix = host_prefix
+
+        super(acbuilder, self).__init__(name, modules_dir, build_dir, install_dir)
+
+    def clean(self):
+        if os.path.exists(self.build_path):
+            os.system("make -C {} uninstall".format(self.build_path))
+        super(acbuilder, self).clean()
+
+    def prepare(self):
+        os.chdir(self.src_path)
+        cmd = ["./build-tools/autogen.sh"]
+        logger.info("preparing {}: {}".format(self.name, " ".join(cmd)))
+        subprocess.check_call(cmd, env=self.env)
+
+        self.configure()
+
+    def configure(self):
+        os.mkdir(self.build_path)
+        os.chdir(self.build_path)
+        cmd = [self.src_path + "/configure",
+               "--prefix=" + self.install_path]
+        if self.host_prefix:
+            cmd.append("--host=" + self.host_prefix)
+        logger.info("configuring {}: {}".format(self.name, " ".join(cmd)))
+        subprocess.check_call(cmd, env=self.env)
+
+    def make(self):
+        cmd = ["make", '-C', self.build_path, "install"]
+        if self.make_verbose:
+            cmd.extend(["V=s"])
+        else:
+            cmd.extend(["-j", str(multiprocessing.cpu_count() + 1)])
+        logger.info("building & installing {}: {}".format(self.name, " ".join(cmd)))
+        subprocess.check_call(cmd, env=self.env)
+
 class mapbuild(object):
     def __init__(self, args):
         if args.verbose: logger.setLevel(logging.DEBUG)
+
         commands = args.commands
-        modules = build_modules if 'all' in args.modules else [m for m in build_modules if m in args.modules]
-        modules_dir = os.path.realpath(args.map_path)
-        build_dir = os.path.realpath(modules_dir + '/build')
-        install_dir = os.path.realpath(build_dir + '/install')
+        _map_modules = map_modules if 'all' in args.modules or 'map' in args.modules else [m for m in map_modules if m in args.modules]
+        _dep_modules = dep_modules if 'all' in args.modules or 'dep' in args.modules else [m for m in dep_modules if m in args.modules]
 
-        if 'distclean' in commands and os.path.exists(build_dir):
-             logger.info("distclean - deleting {}".format(build_dir))
-             shutil.rmtree(build_dir)
+        logger.info("{} {}".format(commands, _map_modules + _dep_modules))
 
-        logger.info("{} {}".format(commands, modules))
-        logger.debug("modules_dir={}, build_dir={}, install_dir={}".format(modules_dir, build_dir, install_dir))
+        # build dep modules
+        for name in _dep_modules:
+            modules_dir = os.path.join(os.path.realpath(args.map_path), "../3rdparty")
+            build_dir = os.path.realpath(modules_dir + '/../multiap/build')
+            install_dir = os.path.realpath(build_dir + '/install')
+
+            if name == 'nng':
+                args.cmake_flags.append("BUILD_SHARED_LIBS=yes")
```

Comment:
Building `nng` is based on `cmake`. Is it needed to be compiled by external toolchain?

Reply:
Maybe in the future, but definitely not now - external_toolchain.cmake is used by map modules only, you can see the implementation in each of the subtrees under external_toolchain.cmake.
We have CMAKE support for cross compiling only map modules, so no need to add support for cross compiling in maptools before we add such support in the modules themselves.
nng might support this but then again we expect nng not to change and already be available in the cross compiled image. So bottom line - no, no need

Reply:
> external_toolchain.cmake is used by map modules only
This is not entirely correct. `CMAKE_TOOLCHAIN_FILE` is standard CMake and should work with any package. What is special for map is that the external_toolchain.cmake is included in the component directories, while normally it is provided by the cross-compilation environment. So indeed, it will not work as is for the non-map components.